lib/phoenix/channel.ex

```diff
@@ -169,9 +169,13 @@ defmodule Phoenix.Channel do
   case of exits. This is the same behaviour as you find in Elixir
   abstractions like `GenServer` and others. Typically speaking, if you
   want to clean something up, it is better to monitor your channel
-  process and do the clean up from another process.  Similar to GenServer,
-  it would also be possible `:trap_exit` to guarantee that `terminate/2`
-  is invoked. This practice is not encouraged though.
+  process and do the clean up from another process. The PID returned
+  by `self()` in `join/3` is the one you want to monitor, you'd
+  normally spawn it in that callback.
```

Comment:
"you'd normally spawn it in that callback." this line not clear to me. Can you please expand what you mean? :)

Reply:
Ah! I mean `join/3`, which was mentioned, maybe word it differently?

Reply:
Oh, maybe "it" is also confusing. The intention is to say: "You'd normally spawn the monitor in `join/3`". Let's rewrite it :).

Reply:
Ah, the misunderstanding is rather with what is being spawned. You mean to spawn a process to monitor it and clean it up. You can also have a central process.

Reply:
Gotcha. To talk with code, the use case I have in mind is something like this (written on the fly here):

```elixir
def join(...) do
  ...
  monitor(socket)
end

def monitor(socket) do
  channel_pid = self()
  spawn(fn ->
    Process.monitor(channel_pid)
    receive do
      {:DOWN, _ref, :process, _monitored_pid, _reason} ->
        on_leave(socket)
    end
  end)
end
```

Where `on_leave()` is a custom function in which yoou log the leave of the user (whose ID is in the socket), and do something for metrics, telemetry, or Prometheus directly, you know.

So perhaps not suggest a particular approach, and leave that detail to the user? Document only the PID?

Which would be the requivalent of that snippet with a central process BTW?

Reply:
I have the same for sockets, but in that case I need a hack to get the transport PID:

```elixir
defoverridable init: 1

@impl true
def init(state) do
  res = {:ok, {_, socket}} = super(state)
  monitor(socket)
  res
end
```

So, in this case, probably that cannot be documented if that hack is using private API.

Reply:
The transport id in the socket is `self()`, no?

The centralized process would be sending a message to a locally registered process with `self()`. Then the process monitors it and call the relevant code when it crashes.

Reply:
> The transport id in the socket is self(), no?

Yes. The transport ID is not `self()` in `connect/3` (you know that, just explaining for the archives), so I need to decorate `init/1`. I believe there is no public interface to get it when the connection was established.

Once you are in `init/1`, chosing `self()` or `socket.transport_pid` is a matter of API usage or respecting the public contract (because the docs do not tell you they are the same).

Reply:
Wait, I actually thought that `self()` in `connect/3` was the transport id. :D I guess that `connect/3` is called before we start the transport.

Btw, why do you need to monitor both transport and channel?

Reply:
> The centralized process would be sending a message to a locally registered process with self(). Then the process monitors it and call the relevant code when it crashes.

In the approach of my snippet, something I like is that the function the monitor runs is isolated in a process. If it does anything weird, I do not need to be defensive about the monitor crashing. However, my skills are very rough, so perhaps there's also a way as simple as that to make a centralized process to be robust.

Reply:
Yup, your approach is totally fine too!

Reply:
Oh, if I spawn when `:DOWN` is received, is the same. And that way the app does not have twice as much processes as joined channels constantly, it only spawns to run the callback, which statistically it is going to be way less simulatenous processes.

Reply:
> Wait, I actually thought that self() in connect/3 was the transport id. :D I guess that connect/3 is called before we start the transport.

@chrismccord told me the reason is that he prefers not to spawn a process until `connect/3` has accepted the connection. So Phoenix needs the return value of the callback in order to decide.

> Btw, why do you need to monitor both transport and channel?

You are going to laugh now :D.

In my app, I only need to monitor socket. I want socket-specific things, like logging disconnections by user, or having a gauge in Prometheus of opened sockets.

I do that at the socket level because that is the actual thing I am interested in. One socket may spawn other 27 channels, so what belongs naturally in the socket, stays attached to the socket.

OK, why I use this approach for sockets and do not provide a doc patch? Because it is a hack using private APIs and I would not know what to propose.

So, why do I provide a doc patch for channels if I am not using this? Because I noticed the lack of public contract about `self()`, so I fired the editor to contribute because you know how I am :D.

Reply:
Oh! I also want to register for how long a user was connected, which is something `:telemetry` does not provide unless I missed it. I am going to assume that the monitor is fired in a timely manner so that the metric won't be exact to the nanosecond, but won't be 5 seconds off if the server is not under a heavy load (this was a practical advice from @michalmuskala).

You see, socket-specific data.

Then channels may have their own specific data too. During the lifetime of a connection people may join and leave and arbitrary number of channels (including none), so what's specifically about the socket, is naturally attached to socket code and processes.

Reply:
I see. There are some things:

1. We can document what `self()` is, but then we are documenting how Elixir behaves, not Phoenix. `self()` is always the current process. It is a matter of knowing what the current process is. :)

2. Even if we document the channel, you are telling me that using the socket is usually better, so this may not have much purpose. :)

Your `init` code could possibly be written like this:

```elixir
def init(arg) do
  monitor(self())
  super(arg)
end
```

`init/1` is public API, so this is OK-ish - but indeed I am not sure how much we want to encourage this. :)

Reply:
> We can document what self() is, but then we are documenting how Elixir behaves, not Phoenix. self() is always the current process. It is a matter of knowing what the current process is. :)

Let me explain how I see it :). What Phoenix needs to tell you is that "the channel process that is going to die when the channel terminates is `self()`". As a user, how does the channel process communicates with your `self()` is something you do not control. That is opaque. You have a callback to do what I want to do, which is `terminate/2`. See? That is public contract. That callback is called in turn by the actual, real GenServer callback, but that is private. Your public interface is `terminate/2` and then you don't need a PID.

But as I said in the feedback, everybody tells you to not use `terminate/2` + traps exits (which by now I am cargo culting), so I need a new contract that guarantees which PID to monitor without revealing the implementation.

Maybe this is all implicit for a seasoned Elixir programmed that perhaps may think something like "there's no other sane/idiomatic way this can be implemented".

But then, looks at what happens with sockets! That is the counterexample. Without looking at the source code, your interface is what the documentation says.

> Even if we document the channel, you are telling me that using the socket is usually better, so this may not have much purpose. :)

No, no! I am saying that in my particular case I am using only a monitor for sockets. But of course there are use cases for channels. Indeed, the first time I saw this approach was in a [SO answer](https://stackoverflow.com/a/33941469/8362448) of @chrismccord for channels.

Reply:
Alternatively, Phoenix would provide a callback `terminate_i_mean_for_real/2`, for sockets, and for channels. That would be the perfect and easy public interface.

Reply:
So would this be a better description? "`join` and  `terminate` are both called from within the channel process. Therefore, `self()` in both callbacks point to the same PID." Then we could expand with examples of the spawn monitor. 

From my understanding, your concerns about self() is related to the process life-cycle. So saying they are both part of the same process life-cycle may help build the necessary mental modal?

Reply:
That is perfect 😀.